### PR TITLE
feat: adding pre-commit hook and makefile installation script

### PR DESCRIPTION
**Pull request by distractedm1nd**
Adds a pre-commit git hook that runs the linter and `go mod tidy`

### CODE REVIEW

**Review by adlerjohn — COMMENTED** [review]
Pre-commit hooks should generally be avoided IMO.

1. They're not portable.
2. They can lead to issues such as https://github.com/trufflesuite/ganache/issues/562
3. They're unexpected, as they mutate code opaquely.

It's better to _verify_ rather than _execute_, i.e. run `go mod tidy` in CI and fail if there's a diff.

**Comment by distractedm1nd** [comment]
@adlerjohn Right, but these are not forced, each contributor has to opt-in to it - prevents `chore: lint` commits and yucky `go mod tidy` commits

**Comment by adlerjohn** [comment]
> each contributor has to opt-in

Do you mean opt-out? The documentation I'm reading says you can opt out with `--no-verify`.

**Comment by distractedm1nd** [comment]
No, I mean opt-in. You only have to install the hooks if you want. They are not installed automatically when downloading the repo (as with husky), thus adding the command in the makefile. Once you have installed them, you can ignore them for single commits with `--no-verify` if you want (for example, checkpointing during refactorings)

**Comment by distractedm1nd** [comment]
@adlerjohn https://github.com/celestiaorg/celestia-app/tree/main/contrib/githooks
Celestia app also has them!

**Comment by adlerjohn** [comment]
Ah got it, thanks.

File: Makefile [hunk 1]
```diff
@@ -11,6 +11,12 @@ help: Makefile
 	@sed -n 's/^##//p' $< | column -t -s ':' |  sed -e 's/^/ /'
 .PHONY: help
 
+## init: Install git-hooks from .githooks directory.
+init-hooks:
```
Comment (Wondertan) [inline Makefile:15]:
```suggestion
## install: Install git-hooks from .githooks directory.
install-hooks:
```

File: .githooks/pre-commit [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+set -eu -o pipefail
+
+if ! command -v golangci-lint &> /dev/null ; then
+    echo "golangci-lint not installed or available in the PATH" >&2
+    echo "please check https://github.com/golangci/golangci-lint" >&2
+    exit 1
+fi
+
+golangci-lint run
```
Comment (Wondertan) [inline .githooks/pre-commit:11]:
```suggestion
make lint 
```

**Review by Wondertan — COMMENTED** [review]
Can we also add `make test` in the precommit hook?

**Comment by distractedm1nd** [comment]
@Wondertan Good question, renaynay and I talked about this and decided to leave testing out of the commit hook, because checkpointing/committing shouldn't necessarily mean all tests pass (and, waiting a long time for tests to pass before the commit goes through gets annoying very quickly)

**Review by Wondertan — APPROVED** [review]
Makes sense

**Comment by vgonkivs** [comment]
I'd suggest:
1) add condition that skips pre-commit hook if go files were not modified(pr with ADRs only)
2) run markdown check in case of 1)

**Review by vgonkivs — CHANGES_REQUESTED** [review]
I'd suggest:

add condition that skips pre-commit hook if go files were not modified(pr with ADRs only)
run markdown check in case of 1)

**Review by Wondertan — APPROVED** [review]

**Comment by adlerjohn** [comment]
Relevant: https://github.com/celestiaorg/celestia-app/pull/672

**Review by vgonkivs — APPROVED** [review]

**Pull request merged**